### PR TITLE
Added OIDC login functionality for HSL

### PR DESCRIPTION
**Pull request by matleppa**
- new login button
- better interworking with LoginPlatforms and ServiceConfiguration
- ability to also remove login settings via Login platforms page
- parameter values from ServiceConfiguration to LoginPlatforms are updated, when Login platforms page is created

Closes #<issue-number>

## Changes
- new login button for HSL
- better interworking with LoginPlatforms and ServiceConfiguration
- ability to also remove login settings via Login platforms page
- parameter values from ServiceConfiguration to LoginPlatforms are updated, when Login platforms page is created

## Developer checklist
This checklist is to be completed by the PR developer:

- [ ] Alternative solutions were compared/discussed before writing code
    - [ ] trade-offs with this solution are considered acceptable
- [ ] Code in this PR adheres to the [project styleguide](CONTRIBUTING.md)
- [ ] This pull request does not decrease project test coverage
- [ ] If the code changes existing database collection(s), migration has been written
- [ ] If UI texts are added or changed, all texts are internationalized

## Reviewer checklist
Reviewed by: @username1

This list is to be completed by the pull request reviewer:
- [ ] Code works as described/expected
- [ ] Code seems to be error free
    - [ ] no browser console errors visible
    - [ ] no server console errors visible
    - [ ] passes CI build
- [ ] Code is written in a way that promotes maintainability
    - [ ] easy to understand
    - [ ] well organized
    - [ ] follows project coding standards and conventions
    - [ ] well documented

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ group: deprecated-2017Q4
 
 language: node_js
 node_js:
-  - '4'
+  - '6'
```
Comment (ilarimikkonen) [inline .travis.yml]: change to version 4 or discard

File: apinf_packages/core/helper_functions/validate_settings.js [hunk 2]
```diff
@@ -33,6 +33,21 @@ export function fiwareSettingsValid (LoginPlatforms) {
   return false;
 }
 
+// Validates OIDC configuration on LoginPlatforms
+export function oidcSettingsValid (LoginPlatforms) {
```
Comment (ilarimikkonen) [inline apinf_packages/core/helper_functions/validate_settings.js]: please change to refer to HSL

File: apinf_packages/login_platforms/server/methods.js [hunk 3]
```diff
@@ -42,36 +38,190 @@ const updateFunctions = {
...
+        serverUrl: settings.hslConfiguration.serverUrl,
+        authorizationEndpoint: settings.hslConfiguration.authorizationEndpoint,
+        tokenEndpoint: settings.hslConfiguration.tokenEndpoint,
+        userinfoEndpoint: settings.hslConfiguration.userinfoEndpoint,
+   //     idTokenWhitelistFields: settings.hslConfiguration.idTokenWhitelistFields || [],
+      });
+    }
+
+    // Return status message
+    return 'OIDC configuration updated successfully';
```
Comment (ilarimikkonen) [inline apinf_packages/login_platforms/server/methods.js]: do we need to localize this?

File: apinf_packages/login_platforms/server/methods.js [hunk 4]
```diff
@@ -42,36 +38,190 @@ const updateFunctions = {
...
     // Try if settings exist
     try {
       const settings = LoginPlatforms.findOne();
+      // Update configuration according to LoginPlatform data
+      status = updateFunctions.saveGithubConfiguration(settings);
+      status = updateFunctions.saveFiwareConfiguration(settings);
+      status = updateFunctions.saveHslConfiguration(settings);
+    } catch (error) {
+      // otherwise show an error
+      const message = `Update gitHub configuration: ${error}`;
```
Comment (ilarimikkonen) [inline apinf_packages/login_platforms/server/methods.js]: not github specific code, please generalize

File: apinf_packages/login_platforms/server/methods.js [hunk 5]
```diff
@@ -42,36 +38,190 @@ const updateFunctions = {
...
+        };
+        // Store fiware values
+        loginParameters.fiwareConfiguration = fiwareConfiguration;
+        changesFound = true;
+      }
+      // Get parameters for OIDC from configuration DB
+      const configHslParameters = ServiceConfiguration.configurations
+      .findOne({ service: 'hsl' });
+
+      // Check OIDC
```
Comment (ilarimikkonen) [inline apinf_packages/login_platforms/server/methods.js]: please change to HSL

File: apinf_packages/oauth_hsl/server/hsl_server.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+      serverUserinfoEndpoint,
+      {
+        headers: {
+          'User-Agent': userAgent,
+          Authorization: `Bearer ${accessToken}`,
+        },
+      }
+    );
+  } catch (err) {
+    const errText = `Failed to fetch userinfo from OIDC ${serverUserinfoEndpoint}: ${err.message}`;
```
Comment (ilarimikkonen) [inline apinf_packages/oauth_hsl/server/hsl_server.js]: OIDC to hsl

File: apinf_packages/oauth_hsl/server/hsl_server.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+          client_secret: OAuth.openSecret(config.secret),
+          /* eslint no-underscore-dangle: ["error", { "allow": ["_redirectUri"] }] */
+          redirect_uri: OAuth._redirectUri('hsl', config),
+          grant_type: 'authorization_code',
+          state: query.state,
+        },
+      }
+    );
+  } catch (err) {
+    const errMsg = `Failed to get token from OIDC ${serverTokenEndpoint}: ${err.message}`;
```
Comment (ilarimikkonen) [inline apinf_packages/oauth_hsl/server/hsl_server.js]: hsl

File: apinf_packages/oauth_hsl/server/hsl_server.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    throw _.extend(new Error(errMsg),
+      { response: err.response });
+  }
+  if (response.data.error) {
+    // if the http response was a json object with an error attribute
+    const errorMsg = `Failed to complete handshake with OIDC
```
Comment (ilarimikkonen) [inline apinf_packages/oauth_hsl/server/hsl_server.js]: hsl

File: apinf_packages/oauth_hsl/server/hsl_server.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+     ${serverTokenEndpoint}: ${response.data.error}`;
+    throw new Error(errorMsg);
+  } else {
+    if (debug) console.log('XXX: getToken response: ', response.data);
+    return response.data;
+  }
+};
+/* Not needed here, at least not now
```
Comment (ilarimikkonen) [inline apinf_packages/oauth_hsl/server/hsl_server.js]: remove commented code

File: apinf_packages/oauth_hsl/server/hsl_server.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+  // decode payload part
+  const decodedIdTokenPayload = Base64.decode(idToken[1]);
+  // JSONify payload in order to be able to refer to it
+  const decodedIdTokenPayloadJSON = JSON.parse(decodedIdTokenPayload);
+  if (debug) console.log('decoded_json', decodedIdTokenPayloadJSON);
+
+  if (debug) console.log('amr=', decodedIdTokenPayloadJSON.amr);
+  // Do not allow login without MFA
+  if (!decodedIdTokenPayloadJSON.amr.includes('mfa')) {
+    throw new Meteor.Error(403, 'User has not MFA in use.');
```
Comment (ilarimikkonen) [inline apinf_packages/oauth_hsl/server/hsl_server.js]: please localise

File: apinf_packages/oauth_hsl/server/hsl_server.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+  serviceData.id = userinfo.sub;
+  serviceData.username = userinfo.preferred_username;
+  serviceData.fullname = userinfo.name;
+  serviceData.accessToken = userinfo.accessToken;
+  serviceData.expiresAt = expiresAt;
+  serviceData.email = userinfo.email;
+  serviceData.amr = decodedIdTokenPayloadJSON.amr;
+
+  /* Not needed here, at least not now
```
Comment (ilarimikkonen) [inline apinf_packages/oauth_hsl/server/hsl_server.js]: please remove commented out code

File: apinf_packages/settings/client/settings.html [hunk 12]
```diff
@@ -86,6 +86,18 @@
             {{/ if }}
           </div>
         </div>
+        <div class="panel panel-default">
```
Comment (ilarimikkonen) [inline apinf_packages/settings/client/settings.html]: please comment as todo for next or so

File: apinf_packages/settings/collection/schema.js [hunk 13]
```diff
@@ -143,6 +143,22 @@ Settings.schema = new SimpleSchema({
       return validation;
     },
   },
+  loginMethods: {
```
Comment (ilarimikkonen) [inline apinf_packages/settings/collection/schema.js]: comment as more work needed with next

File: apinf_packages/login_platforms/server/methods.js (hunk 3 above)
Comment (matleppa) [reply apinf_packages/login_platforms/server/methods.js]: No need, because it is not shown.

**Comment by matleppa** [comment]
This PR is replaced with PR 3612 (HSL login Oauth and MFA)

**Pull request closed without merging**